From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nagarjuna Kristam <nkristam@nvidia.com>
Date: Mon, 14 Nov 2022 00:11:30 +0530
Subject: [PATCH 222/254] NVIDIA: SAUCE: PCI: tegra194: Add PEX_PRSNT GPIO
 support

Read nvidia,pex-prsnt-gpios DT property as a OUT GPIO for EP
and IN GPIO on RP. When EP is started deassert GPIO
to trigger interrupt on RP to indicate hotplug and vice-versa for
hot unplug scenario.

On RP which has nvidia,pex-prsnt-gpios DT property, register a GPIO irq
and configure RP only on GPIO interrupt.

Bug 3868928

Change-Id: Ia1457c7371759712782191a3a68c6fc4bc3c17ba
Signed-off-by: Nagarjuna Kristam <nkristam@nvidia.com>
Reviewed-by: Manikanta Maddireddy <mmaddireddy@nvidia.com>
Reviewed-by: Bitan Biswas <bbiswas@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2897575
Tested-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>

diff --git a/drivers/pci/controller/dwc/pcie-tegra194.c b/drivers/pci/controller/dwc/pcie-tegra194.c
index 0b4dcd805c6ec..ca5f7b8675160 100644
--- a/drivers/pci/controller/dwc/pcie-tegra194.c
+++ b/drivers/pci/controller/dwc/pcie-tegra194.c
@@ -282,10 +282,16 @@ struct tegra_pcie_dw {
 
 	struct dentry *debugfs;
 
+	wait_queue_head_t config_rp_waitq;
+	bool config_rp_done;
+
 	/* Endpoint mode specific */
 	struct gpio_desc *pex_rst_gpiod;
 	struct gpio_desc *pex_refclk_sel_gpiod;
+	struct gpio_desc *pex_prsnt_gpiod;
 	unsigned int pex_rst_irq;
+	unsigned int prsnt_irq;
+	bool perst_irq_enabled;
 	int ep_state;
 	long link_status;
 	struct icc_path *icc_path;
@@ -965,7 +971,14 @@ static int tegra_pcie_dw_start_link(struct dw_pcie *pci)
 	bool retry = true;
 
 	if (pcie->of_data->mode == DW_PCIE_EP_TYPE) {
-		enable_irq(pcie->pex_rst_irq);
+		if (!pcie->perst_irq_enabled) {
+			enable_irq(pcie->pex_rst_irq);
+			pcie->perst_irq_enabled = true;
+		}
+
+		if (pcie->pex_prsnt_gpiod)
+			gpiod_set_value_cansleep(pcie->pex_prsnt_gpiod, 1);
+
 		return 0;
 	}
 
@@ -1057,8 +1070,8 @@ static void tegra_pcie_dw_stop_link(struct dw_pcie *pci)
 {
 	struct tegra_pcie_dw *pcie = to_tegra_pcie(pci);
 
-	if (pcie->of_data->mode == DW_PCIE_EP_TYPE)
-		disable_irq(pcie->pex_rst_irq);
+	if (pcie->pex_prsnt_gpiod)
+		gpiod_set_value_cansleep(pcie->pex_prsnt_gpiod, 0);
 }
 
 static const struct dw_pcie_ops tegra_dw_pcie_ops = {
@@ -1184,6 +1197,17 @@ static int tegra_pcie_dw_parse_dt(struct tegra_pcie_dw *pcie)
 		pcie->enable_srns =
 			of_property_read_bool(np, "nvidia,enable-srns");
 
+	pcie->pex_prsnt_gpiod = devm_gpiod_get_optional(pcie->dev, "nvidia,pex-prsnt",
+			(pcie->of_data->mode == DW_PCIE_RC_TYPE) ? GPIOD_IN : GPIOD_OUT_LOW);
+	if (IS_ERR(pcie->pex_prsnt_gpiod)) {
+		int err = PTR_ERR(pcie->pex_prsnt_gpiod);
+
+		if (err == -EPROBE_DEFER)
+			return err;
+
+		dev_dbg(pcie->dev, "Failed to get PCIe PRSNT GPIO: %d\n", err);
+	}
+
 	if (pcie->of_data->mode == DW_PCIE_RC_TYPE)
 		return 0;
 
@@ -1624,6 +1648,7 @@ static void tegra_pcie_dw_pme_turnoff(struct tegra_pcie_dw *pcie)
 
 static void tegra_pcie_deinit_controller(struct tegra_pcie_dw *pcie)
 {
+	pcie->link_state = false;
 	clk_disable_unprepare(pcie->core_clk_m);
 	dw_pcie_host_deinit(&pcie->pci.pp);
 	tegra_pcie_dw_pme_turnoff(pcie);
@@ -1676,6 +1701,10 @@ static void pex_ep_event_pex_rst_assert(struct tegra_pcie_dw *pcie)
 	if (pcie->ep_state == EP_STATE_DISABLED)
 		return;
 
+	/* Endpoint is going away, assert PRSNT# to mask EP from RP until it is ready link up */
+	if (pcie->pex_prsnt_gpiod)
+		gpiod_set_value_cansleep(pcie->pex_prsnt_gpiod, 0);
+
 	ret = readl_poll_timeout(pcie->appl_base + APPL_DEBUG, val,
 		((val & APPL_DEBUG_LTSSM_STATE_MASK) == LTSSM_STATE_DETECT_QUIET) ||
 		((val & APPL_DEBUG_LTSSM_STATE_MASK) == LTSSM_STATE_DETECT_ACT) ||
@@ -1932,6 +1961,34 @@ static void pex_ep_event_pex_rst_deassert(struct tegra_pcie_dw *pcie)
 	pm_runtime_put_sync(dev);
 }
 
+static irqreturn_t tegra_pcie_prsnt_irq(int irq, void *arg)
+{
+	struct tegra_pcie_dw *pcie = arg;
+	int ret;
+
+	wait_event(pcie->config_rp_waitq, pcie->config_rp_done);
+
+	if (!gpiod_get_value(pcie->pex_prsnt_gpiod)) {
+		if (!pcie->link_state)
+			return IRQ_HANDLED;
+
+		debugfs_remove_recursive(pcie->debugfs);
+		tegra_pcie_deinit_controller(pcie);
+		pm_runtime_put_sync(pcie->dev);
+		pm_runtime_disable(pcie->dev);
+	} else {
+		if (pcie->link_state)
+			return IRQ_HANDLED;
+
+		ret = tegra_pcie_config_rp(pcie);
+		if (ret < 0)
+			dev_err(pcie->dev, "Failed to link up during PCIe hotplug: %d\n",
+				ret);
+	}
+
+	return IRQ_HANDLED;
+}
+
 static irqreturn_t tegra_pcie_ep_pex_rst_irq(int irq, void *arg)
 {
 	struct tegra_pcie_dw *pcie = arg;
@@ -2099,6 +2156,7 @@ static int tegra_pcie_config_ep(struct tegra_pcie_dw *pcie,
 		return -ENOMEM;
 	}
 
+	pcie->perst_irq_enabled = false;
 	irq_set_status_flags(pcie->pex_rst_irq, IRQ_NOAUTOEN);
 
 	pcie->ep_state = EP_STATE_DISABLED;
@@ -2306,7 +2364,49 @@ static int tegra_pcie_dw_probe(struct platform_device *pdev)
 			goto fail;
 		}
 
-		ret = tegra_pcie_config_rp(pcie);
+		init_waitqueue_head(&pcie->config_rp_waitq);
+		pcie->config_rp_done = false;
+
+		if (pcie->pex_prsnt_gpiod) {
+			ret = gpiod_to_irq(pcie->pex_prsnt_gpiod);
+			if (ret < 0) {
+				dev_err(dev, "Failed to get PRSNT IRQ: %d\n",
+					ret);
+				goto fail;
+			}
+			pcie->prsnt_irq = (unsigned int)ret;
+
+			name = devm_kasprintf(dev, GFP_KERNEL,
+					      "tegra_pcie_%u_prsnt_irq",
+					      pcie->cid);
+			if (!name) {
+				dev_err(dev, "Failed to create PRSNT IRQ string\n");
+				ret = -ENOMEM;
+				goto fail;
+			}
+
+			ret = devm_request_threaded_irq(dev,
+							pcie->prsnt_irq,
+							NULL,
+							tegra_pcie_prsnt_irq,
+							IRQF_TRIGGER_RISING |
+							IRQF_TRIGGER_FALLING |
+							IRQF_ONESHOT,
+							name, (void *)pcie);
+			if (ret < 0) {
+				dev_err(dev, "Failed to request IRQ for PRSNT: %d\n",
+					ret);
+				goto fail;
+			}
+			if (gpiod_get_value(pcie->pex_prsnt_gpiod))
+				ret = tegra_pcie_config_rp(pcie);
+		} else {
+			ret = tegra_pcie_config_rp(pcie);
+		}
+
+		/* Now PRST# IRQ thread is ready to execute */
+		pcie->config_rp_done = true;
+		wake_up(&pcie->config_rp_waitq);
 		if (ret && ret != -ENOMEDIUM)
 			goto fail;
 		else
@@ -2352,11 +2452,15 @@ static void tegra_pcie_dw_remove(struct platform_device *pdev)
 		if (!pcie->link_state)
 			return;
 
+		disable_irq(pcie->prsnt_irq);
 		debugfs_remove_recursive(pcie->debugfs);
 		tegra_pcie_deinit_controller(pcie);
 		pm_runtime_put_sync(pcie->dev);
 	} else {
-		disable_irq(pcie->pex_rst_irq);
+		if (pcie->perst_irq_enabled)
+			disable_irq(pcie->pex_rst_irq);
+		if (pcie->pex_prsnt_gpiod)
+			gpiod_set_value_cansleep(pcie->pex_prsnt_gpiod, 0);
 		pex_ep_event_pex_rst_assert(pcie);
 		dw_pcie_ep_deinit(ep);
 	}
@@ -2489,9 +2593,11 @@ static void tegra_pcie_dw_shutdown(struct platform_device *pdev)
 	if (pcie->of_data->mode == DW_PCIE_RC_TYPE) {
 		if (!pcie->link_state)
 			return;
+		if (!pm_runtime_enabled(pcie->dev))
+			return;
 
 		debugfs_remove_recursive(pcie->debugfs);
-
+		disable_irq(pcie->prsnt_irq);
 		disable_irq(pcie->pci.pp.irq);
 		if (IS_ENABLED(CONFIG_PCI_MSI))
 			disable_irq(pcie->pci.pp.msi_irq[0]);
@@ -2500,7 +2606,10 @@ static void tegra_pcie_dw_shutdown(struct platform_device *pdev)
 		tegra_pcie_unconfig_controller(pcie);
 		pm_runtime_put_sync(pcie->dev);
 	} else {
-		disable_irq(pcie->pex_rst_irq);
+		if (pcie->perst_irq_enabled)
+			disable_irq(pcie->pex_rst_irq);
+		if (pcie->pex_prsnt_gpiod)
+			gpiod_set_value_cansleep(pcie->pex_prsnt_gpiod, 0);
 		pex_ep_event_pex_rst_assert(pcie);
 	}
 }
-- 
2.43.0

